### PR TITLE
Tree Navigation: Add visual indicators for items with restricted access

### LINKED ISSUE #20505: Permissions: No error message displayed when user lacks read permission on content node
### Which Umbraco version are you using?

17.0.0-beta

### Bug summary

When a user belongs to a user group that only has "Access to all documents" enabled (without read permissions) and tries to navigate to a content node, the user should receive a notification or message on the content node, like "Not found" or "Not authorized".
However, in version 17, the user only gets a blank page and a console error: `GET https://localhost:44339/umbraco/management/api/v1/document/26abb1b0-c748-4be2-8ad7-be106fdabdee 403 (Forbidden)`

### Specifics

Image

### Steps to reproduce

1. Create a UserGroup with Access to all documents enabled
2. Create TestUser with the UserGroup
3. Create a content node.
4. Log in to the TestUser 
5. Navigate to the created content node
6. See a blank/empty content node

### Expected result / actual result

Expected: Some text should be displayed indicating "Not found" or "Not authorized", as was the case in previous versions.
Actual: A blank/empty content pane is displayed.

### PR DESCRIPTION
**Pull request by iOvergaard**
## Summary

Adds visual styling and behavior for tree items where users can see but cannot navigate to content (ancestors of their start nodes). When a user's start node is nested deep in the tree, they need visual feedback showing which ancestor nodes they don't have access to.

This implements the UI handling for Feature AB#63060 "Handle Start Nodes"

This PR also fixes #20505 by ensuring the workspace shows "Access denied" when users navigate to forbidden non-variant documents via direct URLs.

### User Scenario
When a user has a start node at a grandchild level (e.g., `/Root/Child/Grandchild`), they need to:
1. See the tree structure leading to their start node
2. Understand which ancestor nodes they cannot access
3. Expand those ancestors to navigate to their accessible content

### UI Changes

**Visual Styling for noAccess Items:**
- Native disabled styling from `uui-menu-item` component
- Items are visible and expandable but cannot be clicked to navigate

image

**Implementation Details:**
- Base class (`tree-item-element-base.ts`) provides `_noAccess` property
- When `_noAccess=true`:
  - `disabled` attribute is set on `uui-menu-item` (prevents clicks and keyboard navigation)
  - `href` is removed (prevents navigation)
  - Expand button remains functional (users can navigate through ancestors)
- Child classes (document, media) subscribe to context to set `_noAccess`

### Tree Type Support
- ✅ Document tree (`document-tree-item.element.ts`) - subscribes to `noAccess` from entity context
- ✅ Media tree (`media-tree-item.element.ts`) - same pattern for consistency

## Bug Fix #20505: "Access Denied" Not Showing for Non-Variant Documents

**Problem:** When users received direct links to **non-variant documents** (single language) they don't have access to, navigating to that URL would fail to show the "Access denied" message. This worked correctly for variant documents (multiple languages/cultures) but not for invariant content.

**Root Cause:** In `document-workspace-editor.element.ts`, the `#generateRoutes()` method had an early return for non-variant documents that skipped adding the forbidden route:

```typescript
#generateRoutes() {
  if (!this.#variants || this.#variants.length === 0 || !this.#appCulture) {
    this._routes = [];
    return; // ❌ Returned early without adding forbidden route
  }
  
  // Generate routes for variant documents...
  this.#ensureForbiddenRoute(routes); // ✅ Only called for variant documents
}
```

**Fix:** Call `#ensureForbiddenRoute()` before the early return to ensure both variant and non-variant documents get the catch-all forbidden route:

```typescript
#generateRoutes() {
  if (!this.#variants || this.#variants.length === 0 || !this.#appCulture) {
    this._routes = [];
    this.#ensureForbiddenRoute(this._routes); // ✅ Now handles non-variant documents
    return;
  }
  
  // Generate routes for variant documents...
  this.#ensureForbiddenRoute(routes); // ✅ Still works for variant documents
}
```

The `#ensureForbiddenRoute()` method adds a catch-all route (`**`) that displays either `UmbRouteForbiddenElement` (access denied) or `UmbRouteNotFoundElement` (404) based on the `forbidden` context state.

**User Impact:**
- ✅ "Access denied" now displays correctly for non-variant documents when accessed via direct URLs
- ✅ Variant documents continue to work as before
- ✅ Proper distinction between "forbidden" (permission issue) and "not found" (document doesn't exist)

**Test Coverage:**
- Re-enabled previously skipped E2E test: `ContentStartNodes.spec.ts` - "can see parent of start node but not access it"
- Test verifies that clicking on a noAccess parent shows "Access denied" in the workspace

## Backend Bug Fix: Path Comparison False Matches

**Problem:** Node IDs that are numeric prefixes of other IDs (100 vs 1001, 10 vs 100) caused false matches in access logic.

**Impact:** Tree items were incorrectly marked as accessible or restricted.

**Example:**
- User has start node at ID 1001
- Tree item with ID 100 would incorrectly match
- Reason: `"-1,100".StartsWith("-1,10")` returns `true`

**Fix:** Add trailing commas to paths before comparison (`UserStartNodeEntitiesService.cs` lines 146, 226):

```csharp
// Before: "-1,100".StartsWith("-1,10") = true ❌
// After: "-1,100,".StartsWith("-1,10,") = false ✅

private static bool IsDescendantOrSelf(IEntitySlim child, string[] userStartNodePaths)
    => userStartNodePaths.Any(path => $"{child.Path},".StartsWith($"{path},"));
```

## Architecture Decisions

**Leveraging Native Platform Behavior**

Uses `uui-menu-item`'s native `disabled` attribute instead of custom event handlers:
- ✅ Simpler, more declarative code
- ✅ Better performance (no event listeners)
- ✅ Automatic click and keyboard prevention
- ✅ Expand button functionality preserved (separate event)
- ✅ Centralized in base class
- ✅ Native disabled styling from the component library

**Opt-In Pattern**

Base class provides infrastructure, child classes explicitly subscribe to enable it:
- ✅ Non-breaking change for tree types that don't need access restrictions
- ✅ Clear ownership of feature activation
- ✅ Flexibility for future tree implementations

**Trailing Comma Technique**

Simple string manipulation for path comparison:
- ✅ Minimal performance impact
- ✅ Easy to understand and maintain
- ✅ No additional dependencies
- ✅ Handles all edge cases correctly

## Test Coverage

**Unit Tests (19 tests) - New**

File: UserStartNodePathComparisonTests.cs
- Documents the path comparison bug and fix
- Covers edge cases: numeric prefixes (1/10/100/1001), deep paths, self-matching
- Includes "bug demonstration" tests showing the problem without the fix

**Integration Tests (9 tests) - Passing**

File: UserStartNodeEntitiesServiceTests.cs
- Verifies correct noAccess flags on tree items
- Tests various start node configurations
- Pagination and filtering scenarios

**E2E Tests - Updated**

- ContentStartNodes.spec.ts - Document tree noAccess behavior (previously skipped test now re-enabled)
- MediaStartNodes.spec.ts - Media tree noAccess behavior
- Tests expansion of noAccess ancestors and navigation to accessible children

**Breaking Changes**

None. This is an additive feature with opt-in activation.

**Dependencies**

No new dependencies added.

---
**Testing**: All unit tests (19/19), integration tests (9/9), and E2E tests pass.

**Manual Verification**: Tested in browser with users at various start node levels (root, child, grandchild).

### CODE REVIEW

File: tests/Umbraco.Tests.AcceptanceTest/tests/DefaultConfig/Users/Permissions/User/MediaStartNodes.spec.ts [hunk 1]
```diff
@@ -50,30 +81,102 @@ test('can see root media start node and children', async ({umbracoApi, umbracoUi
...
+    testUser.password
+  );
   await umbracoUi.goToBackOffice();
 
   // Act
   await umbracoUi.user.goToSection(ConstantHelper.sections.media, false);
 
   // Assert
+
+  // Get initial URL (should be on media section)
```
[the remark this answers is not in the document]
Comment (iOvergaard) [reply tests/Umbraco.Tests.AcceptanceTest/tests/DefaultConfig/Users/Permissions/User/MediaStartNodes.spec.ts:110]: Great, I fixed it!

File: tests/Umbraco.Tests.AcceptanceTest/tests/DefaultConfig/Users/Permissions/User/ContentStartNodes.spec.ts [hunk 2]
```diff
@@ -1,50 +1,94 @@
...
+  umbracoUi,
+}) => {
   // Arrange
-  await umbracoApi.user.setUserPermissions(testUser.name, testUser.email, testUser.password, userGroupId, [rootDocumentId]);
-  testUserCookieAndToken = await umbracoApi.user.loginToUser(testUser.name, testUser.email, testUser.password);
+  await umbracoApi.user.setUserPermissions(
+    testUser.name,
+    testUser.email,
+    testUser.password,
+    userGroupId,
```
[the remark this answers is not in the document]
Comment (iOvergaard) [reply tests/Umbraco.Tests.AcceptanceTest/tests/DefaultConfig/Users/Permissions/User/ContentStartNodes.spec.ts:84]: Yeah, but the tests are just made like that throughout the AcceptanceTest project.

File: tests/Umbraco.Tests.AcceptanceTest/tests/DefaultConfig/Users/Permissions/User/MediaStartNodes.spec.ts [hunk 3]
```diff
@@ -1,43 +1,74 @@
...
+  umbracoUi,
+}) => {
   // Arrange
-  await umbracoApi.user.setUserPermissions(testUser.name, testUser.email, testUser.password, userGroupId, [], false, [rootFolderId]);
-  testUserCookieAndToken = await umbracoApi.user.loginToUser(testUser.name, testUser.email, testUser.password);
+  await umbracoApi.user.setUserPermissions(
+    testUser.name,
+    testUser.email,
+    testUser.password,
+    userGroupId,
```
[the remark this answers is not in the document]
Comment (iOvergaard) [reply tests/Umbraco.Tests.AcceptanceTest/tests/DefaultConfig/Users/Permissions/User/MediaStartNodes.spec.ts:62]: Yeah, but the tests are just made like that throughout the AcceptanceTest project.

File: src/Umbraco.Web.UI.Client/src/packages/documents/documents/workspace/document-workspace-editor.element.ts [hunk 4]
```diff
@@ -143,15 +144,24 @@ export class UmbDocumentWorkspaceEditorElement extends UmbLitElement {
 			});
 		}
 
+		this.#ensureForbiddenRoute(routes);
+
+		this._routes = routes;
```
[the remark this answers is not in the document]
Comment (iOvergaard) [reply src/Umbraco.Web.UI.Client/src/packages/documents/documents/workspace/document-workspace-editor.element.ts:149]: Great catch - fixed!

File: src/Umbraco.Web.UI.Client/src/packages/media/media/tree/tree-item/media-tree-item.element.ts [hunk 5]
```diff
@@ -58,7 +29,9 @@ export class UmbMediaTreeItemElement extends UmbTreeItemElementBase<UmbMediaTree
 	}
 
 	override renderLabel() {
-		return html`<span id="label" slot="label">${this._item?.variants[0].name}</span> `;
+		return html`<span id="label" slot="label" class=${classMap({ noAccess: this._noAccess })}>
+			${this._item?.variants[0].name}
+		</span> `;
```
[the remark this answers is not in the document]
Comment (iOvergaard) [reply src/Umbraco.Web.UI.Client/src/packages/media/media/tree/tree-item/media-tree-item.element.ts:34]: I know, but the draft class is also there and it gives the developer an indicator as to why a node is formatted in a certain way, e.g. `draft` for non-published documents (pre-existing) and now `noAccess` for restricted nodes.

File: src/Umbraco.Web.UI.Client/src/packages/documents/documents/tree/tree-item/document-tree-item.element.ts [hunk 6]
```diff
@@ -85,7 +68,9 @@ export class UmbDocumentTreeItemElement extends UmbTreeItemElementBase<
 	};
 
 	override renderLabel() {
-		return html`<span id="label" slot="label" class=${classMap({ draft: this._isDraft })}>${this._name}</span> `;
+		return html`<span id="label" slot="label" class=${classMap({ draft: this._isDraft, noAccess: this._noAccess })}>
+			${this._name}
+		</span> `;
```
[the remark this answers is not in the document]
Comment (iOvergaard) [reply src/Umbraco.Web.UI.Client/src/packages/documents/documents/tree/tree-item/document-tree-item.element.ts:73]: I know, but the draft class is also there and it gives the developer an indicator as to why a node is formatted in a certain way, e.g. draft for non-published documents (pre-existing) and now noAccess for restricted nodes.

File: src/Umbraco.Cms.Api.Management/Services/Entities/UserStartNodeEntitiesService.cs [hunk 7]
```diff
@@ -220,5 +221,7 @@ public IEnumerable<UserAccessEntity> UserAccessEntities(IEnumerable<IEntitySlim>
         => entities.Select(entity => new UserAccessEntity(entity, IsDescendantOrSelf(entity, userStartNodePaths))).ToArray();
 
     private static bool IsDescendantOrSelf(IEntitySlim child, string[] userStartNodePaths)
-        => userStartNodePaths.Any(path => child.Path.StartsWith(path));
+        // Note: Add trailing commas to both paths to prevent false matches (e.g., path "-1,100" should not match "-1,1001")
+        // This matches the pattern used in lines 92 and 192 of this file
```
[the remark this answers is not in the document]
Comment (iOvergaard) [reply src/Umbraco.Cms.Api.Management/Services/Entities/UserStartNodeEntitiesService.cs:225]: Fixed, thanks!

File: src/Umbraco.Web.UI.Client/src/packages/core/tree/tree-item/tree-item-base/tree-item-element-base.ts
Comment (nielslyngsoe) [inline src/Umbraco.Web.UI.Client/src/packages/core/tree/tree-item/tree-item-base/tree-item-element-base.ts]: A though from here:
I think you can avoid the prevention of events by setting disabled and not set the href when in noAccess mode.

But you have to see how that works in relation to pickers, and especially picker with ignorer start nodes, cause in that case you should still be able to select those nodes.

You may just need to incorporate the _isSelectableContext into the logic that sets disabled when in noaccess mode.

I hope those thoughts makes sense,  this should give better user feedback as the item will appear not clickable and we can avoid the complixity of these event-prevent-handlers :-)

Comment (iOvergaard) [reply src/Umbraco.Web.UI.Client/src/packages/core/tree/tree-item/tree-item-base/tree-item-element-base.ts]: Great catch, that has been implemented just now.
